src/test/util/net.h

```diff
@@ -133,110 +136,228 @@ constexpr auto ALL_NETWORKS = std::array{
+
+    int Connect(const sockaddr*, socklen_t) const override;
+
+    int Bind(const sockaddr*, socklen_t) const override;
+
+    int Listen(int) const override;
+
+    std::unique_ptr<Sock> Accept(sockaddr* addr, socklen_t* addr_len) const override;
+
+    int GetSockOpt(int level, int opt_name, void* opt_val, socklen_t* opt_len) const override;
+
+    int SetSockOpt(int, int, const void*, socklen_t) const override;
+
+    int GetSockName(sockaddr* name, socklen_t* name_len) const override;
+
+    bool SetNonBlocking() const override;
+
+    bool IsSelectable() const override;
+
+    bool Wait(std::chrono::milliseconds timeout,
+              Event requested,
+              Event* occurred = nullptr) const override;
+
+    bool WaitMany(std::chrono::milliseconds timeout, EventsPerSock& events_per_sock) const override;
+
+private:
+    ZeroSock& operator=(Sock&& other) override;
+};
+
 /**
  * A mocked Sock alternative that returns a statically contained data upon read and succeeds
  * and ignores all writes. The data to be returned is given to the constructor and when it is
  * exhausted an EOF is returned by further reads.
  */
-class StaticContentsSock : public Sock
+class StaticContentsSock : public ZeroSock
 {
 public:
-    explicit StaticContentsSock(const std::string& contents)
-        : Sock{INVALID_SOCKET},
-          m_contents{contents}
-    {
-    }
+    explicit StaticContentsSock(const std::string& contents);
 
-    ~StaticContentsSock() override { m_socket = INVALID_SOCKET; }
+    /**
+     * Return parts of the contents that was provided at construction until it is exhausted
+     * and then return 0 (EOF).
+     */
+    ssize_t Recv(void* buf, size_t len, int flags) const override;
 
-    StaticContentsSock& operator=(Sock&& other) override
+    bool IsConnected(std::string&) const override
     {
-        assert(false && "Move of Sock into MockSock not allowed.");
-        return *this;
+        return true;
     }
 
-    ssize_t Send(const void*, size_t len, int) const override { return len; }
+private:
+    StaticContentsSock& operator=(Sock&& other) override;
 
-    ssize_t Recv(void* buf, size_t len, int flags) const override
+    const std::string m_contents;
+    mutable size_t m_consumed{0};
+};
+
+/**
+ * A mocked Sock alternative that allows providing the data to be returned by Recv()
+ * and inspecting the data that has been supplied to Send().
+ */
+class DynSock : public ZeroSock
+{
+public:
+    /**
+     * Unidirectional bytes or CNetMessage queue (FIFO).
+     */
+    class Pipe
     {
-        const size_t consume_bytes{std::min(len, m_contents.size() - m_consumed)};
-        std::memcpy(buf, m_contents.data() + m_consumed, consume_bytes);
-        if ((flags & MSG_PEEK) == 0) {
-            m_consumed += consume_bytes;
-        }
-        return consume_bytes;
-    }
+    public:
+        /**
+         * Get bytes and remove them from the pipe.
+         * @param[in] buf Destination to write bytes to.
```

Comment:
nit: this is an `out` parameter, pointed by @pinheadmz in https://github.com/Sjors/bitcoin/pull/64#discussion_r1878686775
Diff that I applied locally and will be in next retouch (if there is next retouch):

```diff
    -+         * @param[in] buf Destination to write bytes to.
    ++         * @param[out] buf Destination to write bytes to.
```